File: 005Serialization_and_deserialization_of_binary_trees/005Serialization_and_deserialization_of_binary_trees.py
# Definition for a binary tree node.
import pdb
from queue import Queue

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        add=""
        if root == None:
            return "N,"
        add += (str(root.val))
        add += ','
        add += self.serialize(root.left)
        add += self.serialize(root.right)
        return add
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        result_data = data.split(',')
        q = Queue()
        for i in result_data:
            q.put(i)
        result = self.from_queue(q)

        return result

    def from_queue(self,result_queue):
        root = result_queue.get()
        if root == 'N':
            return None
        result = TreeNode(int(root))
        result.left = self.from_queue(result_queue)
        result.right = self.from_queue(result_queue)
        return result
        
root = TreeNode(1)
child_1 = TreeNode(2)
child_2 = TreeNode(3)
root.left = child_1
root.right = child_2
child_3 = TreeNode(4)
child_4 = TreeNode(5)
child_2.left = child_3
child_2.right = child_4

# Your Codec object will be instantiated and called as such:
codec = Codec()
#pdb.set_trace()
result = codec.serialize(root)
print(result)
head = codec.deserialize(result)
#pdb.set_trace()
print(codec.serialize(head))

File: 005Serialization_and_deserialization_of_binary_trees/005Serialization_and_deserialization_of_binary_trees.cpp
#include<string>
#include<iostream>
#include<queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val = 0;
    TreeNode* left = NULL;
    TreeNode* right = NULL;
    //TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:

    // 这里使用string储存，是因为需求为转换为字符串!
    // Encodes a tree to a single string.
    // 没用实际意义，还请看deserialize实现string按','分开的代码，这一部分没有单独做成函数
    string serialize(TreeNode* root) {
        string add{};
        if (root == NULL)
            return string("N,");
        add += to_string(root->val);
        add += ",";
        add += serialize(root->left);
        //add += to_string(root->val);
        add += serialize(root->right);
        //add += to_string(root->val);
        return add;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        string* result_data = new string[data.size()];
        int begin = 0;
        int l = 0;
        int count = 0;
        if (data.size() == 0)
            return NULL;
        for (int i = 0;i < data.size();i++)
        {
            if (data[i] == ',')
            {
                string s(data, begin, i - begin);
                result_data[l++] = s;
                begin = i + 1;
                count++;
            }
        }
        // 以上，*result_data储存按','区分的各个val的值
        
        queue<string> result_queue;
        for (int i = 0;i < count;i++)
            result_queue.push(result_data[i]);

        TreeNode* result = from_queue(result_queue);
        return result;
    }

    TreeNode* from_queue(queue<string>& result_queue) {
        string root = result_queue.front();
        result_queue.pop();
        if (root == string("N"))
            return NULL;
        TreeNode* result = new TreeNode;
        result->val = stoi(root);
        result->left = from_queue(result_queue);
        result->right = from_queue(result_queue);
        return result;
    }
};

int main() {
    TreeNode* root = new TreeNode;
    root->val = 1;
    TreeNode* child_1 = new TreeNode;
    child_1->val = 2;
    root->left = child_1;
    TreeNode* child_2 = new TreeNode;
    child_2->val = 3;
    root->right = child_2;
    TreeNode* child_3 = new TreeNode;
    child_3->val = 4;
    child_2->left = child_3;
    TreeNode* child_4 = new TreeNode;
    child_4->val = 5;
    child_2->right = child_4;
    Codec codec;
    string root_code= codec.serialize(root);
    cout << root_code << endl;
    cout << codec.serialize(codec.deserialize(root_code)) << endl;
    return 0;
}
// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));